mixpanel_oauth_to_redshift.go

```diff
@@ -0,0 +1,177 @@
+)
+
+var (
+	aws_access_key_id     = os.Getenv("AWS_ACCESS_KEY_ID")
+	aws_secret_access_key = os.Getenv("AWS_SECRET_ACCESS_KEY")
+	aws_region            = os.Getenv("AWS_REGION")
+	mixpanel_api_key      = os.Getenv("MIXPANEL_API_KEY")
+	mixpanel_api_secret   = os.Getenv("MIXPANEL_API_SECRET")
+	redshift_host         = os.Getenv("AWS_REDSHIFT_HOST")
+	redshift_port         = os.Getenv("AWS_REDSHIFT_PORT")
+	redshift_database     = os.Getenv("AWS_REDSHIFT_DB_NAME")
+	redshift_user         = os.Getenv("AWS_REDSHIFT_USER")
+	redshift_password     = os.Getenv("AWS_REDSHIFT_PASSWORD")
+	jsonpaths_file        = flag.String("jsonpaths_file", "s3://clever-mixpanel/OAuth.json",
+		"s3 file with jsonpaths data.")
+	mixpanel_export_date = flag.String("mixpanel_export_date",
+		time.Now().AddDate(0, 0, -1).Format("2006-01-02"),
+		"Date in YYYY-MM-DD format. Defaults to yesterday.")
+	mixpanel_export_dir = flag.String("mixpanel_export_dir", "s3://clever-mixpanel/OAuth",
+		"Directory to store the exported mixpanel data.")
+	mixpanel_export_request_timeout = flag.Duration(
+		"mixpanel_export_request_timeout", 10*time.Minute,
+		"Timeout value to be used while making a mixpanel export request. Defaults to 10 mins.")
+	redshift_table       = flag.String("redshift_table", "mixpanel_oauth", "Name of the redshift table.")
+	export_from_mixpanel = flag.Bool("export_from_mixpanel", true, "Whether to export from mixpanel.")
+	copy_to_redshift     = flag.Bool("copy_to_redshift", true, "Whether to copy to redshift.")
+)
+
+type MixpanelExport struct {
+	endpoint   string
+	version    string
+	api_key    string
+	api_secret string
+}
+
+func NewMixpanelExport() *MixpanelExport {
+	m := MixpanelExport{"https://data.mixpanel.com/api", "2.0", mixpanel_api_key, mixpanel_api_secret}
+	return &m
+}
+
+// json.Marshal wraps a string with quotes which breaks URL params in API calls. This function fixes
+// that problem.
+func StringOrJsonMarshal(value interface{}) (string, error) {
+	str_value, ok := value.(string)
+	if !ok {
+		json_value, err := json.Marshal(value)
+		if err != nil {
+			return "", err
+		}
+		str_value = string(json_value)
+	}
+	return str_value, nil
+}
+
+func (m *MixpanelExport) ComputeSig(params map[string]interface{}) (string, error) {
+	var keys []string
+
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	joined_args := ""
+	for _, k := range keys {
+		joined_args += k
+		value, err := StringOrJsonMarshal(params[k])
+		if err != nil {
+			return "", err
+		}
+		joined_args += fmt.Sprintf("=%s", value)
+	}
+
+	hash := md5.Sum([]byte(joined_args + m.api_secret))
+	return hex.EncodeToString(hash[:]), nil
+}
+
+func (m *MixpanelExport) Request(method string, params map[string]interface{}) ([]byte, error) {
+	params["api_key"] = m.api_key
+	expire_time := time.Now().Add(*mixpanel_export_request_timeout)
+	params["expire"] = expire_time.Unix()
+	params["format"] = "json"
+	delete(params, "sig")
```

Comment:
Might deserve a comment on why you're deleting this. (would it just be user error to pass "sig" as a param?)